zebra-chain/src/orchard/sinsemilla.rs

```diff
@@ -0,0 +1,150 @@
+//! Sinsemilla hash functions and helpers.
+
+use bitvec::prelude::*;
+
+use halo2::{
+    arithmetic::{CurveAffine, CurveExt},
+    pasta::pallas,
+};
+
+/// [Hash Extractor for Pallas][concreteextractorpallas]
+///
+/// P → B^[l^Orchard_Merkle]
+///
+/// [concreteextractorpallas]: https://zips.z.cash/protocol/nu5.pdf#concreteextractorpallas
+pub fn extract_p(point: pallas::Point) -> pallas::Base {
+    match pallas::Affine::from(point).get_xy().into() {
+        // If Some, it's not the identity.
+        Some((x, _)) => x,
+        _ => pallas::Base::zero(),
+    }
+}
+
+/// GroupHash into Pallas, aka _GroupHash^P_
+///
+/// Produces a random point in the Pallas curve.  The first input element acts
+/// as a domain separator to distinguish uses of the group hash for different
+/// purposes; the second input element is the message.
+///
+/// https://zips.z.cash/protocol/nu5.pdf#concretegrouphashpallasandvesta
+#[allow(non_snake_case)]
+pub fn pallas_group_hash(D: &[u8], M: &[u8]) -> pallas::Point {
+    let domain_separator = std::str::from_utf8(D).unwrap();
+
+    pallas::Point::hash_to_curve(domain_separator)(M)
+}
+
+/// Q(D) := GroupHash^P(︀“z.cash:SinsemillaQ”, D)
+///
+/// https://zips.z.cash/protocol/nu5.pdf#concretesinsemillahash
+#[allow(non_snake_case)]
+fn Q(D: &[u8]) -> pallas::Point {
+    pallas_group_hash(b"z.cash:SinsemillaQ", D)
+}
+
+/// S(j) := GroupHash^P(︀“z.cash:SinsemillaS”, LEBS2OSP32(I2LEBSP32(j)))
+///
+/// S: {0 .. 2^k - 1} -> P^*, aka 10 bits hashed into the group
+///
+/// https://zips.z.cash/protocol/nu5.pdf#concretesinsemillahash
+#[allow(non_snake_case)]
+fn S(j: &u16) -> pallas::Point {
+    // The value of j is a 10-bit value, therefore must never exceed 2^10 in
+    // value.
+    assert!(j < &1024u16);
+
+    pallas_group_hash(b"z.cash:SinsemillaS", &j.to_le_bytes())
+}
+
+/// "...an algebraic hash function with collision resistance (for fixed input
+/// length) derived from assumed hardness of the Discrete Logarithm Problem on
+/// the Pallas curve."
+///
+/// SinsemillaHash is used in the definitions of Sinsemilla commitments and of
+/// the Sinsemilla hash for the Orchard incremental Merkle tree (§ 5.4.1.3
+/// ‘MerkleCRH^Orchard Hash Function’).
+///
+/// https://zips.z.cash/protocol/nu5.pdf#concretesinsemillahash
+///
+/// # Panics
+///
+/// If `M` is greater than `k*c = 2530` bits.
+#[allow(non_snake_case)]
+pub fn sinsemilla_hash_to_point(D: &[u8], M: &BitVec<Lsb0, u8>) -> pallas::Point {
```

Comment:
General comment for this file:

Can we type alias `BitVec<Lsb0, u8>` to `SinsemillaMessage`?
It would make the code a lot easier to read (and use externally).

Reply:
I'm not sure that's the best name, as these are various slices of bits and bytes, but I may make some type aliases as I integrate the test vectors